sphinx_immaterial/nav_adapt.py

```diff
@@ -359,20 +375,141 @@ def _collapse_children_not_on_same_page(entry: MkdocsNavEntry) -> MkdocsNavEntry
+        _add_domain_info_to_toc(app, global_toc, fake_pagename)
+        self.entries = global_toc
+        self.url_map = _build_toc_index(global_toc)
+
+
+def _get_cached_globaltoc_info(app: sphinx.application.Sphinx) -> CachedTocInfo:
+    """Obtains the cached global TOC, generating it if necessary."""
+    key = "sphinx_immaterial_global_toc_cache"
+    data = getattr(app.env, key, None)
+    if data is not None:
+        return data
+    data = CachedTocInfo(app)
+    setattr(app.env, key, data)
+    return data
+
+
+def _get_ancestor_keys(keys: Iterable[TocEntryKey]) -> Set[TocEntryKey]:
+    ancestors = set()
+    for key in keys:
+        while key not in ancestors:
+            ancestors.add(key)
+            key = key[:-1]
+    return ancestors
+
+
+def _get_global_toc(app: sphinx.application.Sphinx, pagename: str, collapse: bool):
+    """Obtains the global TOC for a given page."""
+    cached_data = _get_cached_globaltoc_info(app)
+    url = app.builder.get_target_uri(pagename)
+    keys = set(cached_data.url_map[url])
+    ancestors = _get_ancestor_keys(keys)
+
+    fake_pagename = ""
+
+    fake_page_url = app.builder.get_target_uri(fake_pagename)
+
+    real_page_url = app.builder.get_target_uri(pagename)
+
+    def _make_toc_for_page(key: TocEntryKey, children: List[MkdocsNavEntry]):
+        children = list(children)
+        for i, child in enumerate(children):
+            child_key = key + (i,)
+            child = children[i] = copy.copy(child)
+            if child.url is not None:
+                root_relative_url = urllib.parse.urljoin(fake_page_url, child.url)
+                uri = urllib.parse.urlparse(root_relative_url)
+                if not uri.netloc:
+                    child.url = sphinx.util.osutil.relative_uri(
+                        real_page_url, root_relative_url
+                    )
+                    if uri.fragment:
+                        child.url += f"#{uri.fragment}"
+            in_ancestors = child_key in ancestors
+            if in_ancestors:
+                child.active = True
+                if child_key in keys:
+                    child.current = True
+            if in_ancestors or child.caption_only:
+                child.children = _make_toc_for_page(child_key, child.children)
+            else:
+                child.children = []
+        return children
+
+    return _make_toc_for_page((), cached_data.entries)
+
+
 def _get_mkdocs_tocs(
     app: sphinx.application.Sphinx, pagename: str, duplicate_local_toc: bool
 ) -> Tuple[List[MkdocsNavEntry], List[MkdocsNavEntry]]:
     theme_options = app.config["html_theme_options"]
-    global_toc_node = sphinx.environment.adapters.toctree.TocTree(
-        app.env
-    ).get_toctree_for(
-        pagename,
-        app.builder,
+    global_toc = _get_global_toc(
+        app=app,
+        pagename=pagename,
         collapse=theme_options.get("globaltoc_collapse", False),
```

Comment:
I think these are positional args, so there's no need to use kwarg syntax here.

Reply:
In Python there isn't a clear separation of positional vs keyword args, in that most of the time people don't use `/` or `*` to require one or the other.  Generally I prefer keyword args since it leaves less room for confusion, although granted in this case it doesn't add much.